### PR TITLE
Intermediate.ATX.2017Q1.Clayton.Berger

### PR DESCRIPTION
**Pull request by HippErger**
# Checkpoint Rubric

This is the rubric that your instructor will use to grade your checkpoints. Please do not edit. 

## Checkpoint 1: Towers of Hanoi Terminal App
- [x] All tests passed: 55 points
- [x] Properly commented code: 15 points
- [ ] Properly indented code: 15 points
- [x] Demonstrated effective use of JavaScript: 15 points

Specific Comments:

*Comments may also be made on the file itself using github's commenting tools.*

## Checkpoint 2: Towers of Hanoi GUI App
- [x] All tests passed: 55 points
- [x] Properly commented code: 15 points
- [x] Properly indented code: 15 points
- [x] Demonstrated effective use of jQuery: 15 points

Specific Comments:

*Comments may also be made on the file itself using github's commenting tools.*

## Checkpoint 3: Final Project Gist Blog
- [ ] Effective use of AJAX: 20 points
- [ ] Effective use of loops: 20 points
- [ ] Effective use of github API: 20 points
- [ ] Style and design: 10 points
- [ ] Properly commented code: 15 points
- [ ] Properly indented code: 15 points

Specific Comments:

*Comments may also be made on the file itself using github's commenting tools.*

### CODE REVIEW

File: apps/01PigLatin.js [hunk 1]
```diff
@@ -8,13 +8,23 @@ var rl = readline.createInterface({
 });
 
 
-function pigLatin(word) {
+function pigLatin(word){
```
Comment (titobyte) [inline apps/01PigLatin.js:11]: Add a space after "(word)" before the curly braces.

File: apps/02RockPaperScissors.js [hunk 2]
```diff
@@ -9,8 +9,65 @@ var rl = readline.createInterface({
...
+    if (hand2 === 'paper') {
+      return "Hand One Wins!"
+    }
+    
+   }
+//if a player types in something that isn't one of the three words they will be prompted to try again.
+  if (hand1 !== 'rock' && hand1 !== 'paper' && hand1 !== 'scissors') {
+    return "Invalid Entry, Player 1, Put useable input!"
+  }
+
```
Comment (titobyte) [inline apps/02RockPaperScissors.js]: Nice use of &&

File: apps/03TicTacToe.js [hunk 3]
```diff
@@ -22,25 +30,88 @@ function printBoard() {
...
 
+
+
+
 function ticTacToe(row, column) {
-  // Your code here
+  // Calls takes input from user and places it into row and column in the original board. 
+  board[row][column] = playerTurn;
+  // the next lines check for wins in the rows, columns and diagonals. 
+  if (checkForWin() === true) {
```
Comment (pieroapretto) [inline apps/03TicTacToe.js:88]: If lines 89-103 only run when `checkForWin() === true`, you should place this logic inside that corresponding function.

File: apps/04TowersOfHanoi.js [hunk 4]
```diff
@@ -19,27 +19,98 @@ function printStacks() {
...
+    var lengthOfEndStack = stacks[endStack].length;
+
+    // Check to see if the stack being pulled from, startStack, has tokens/donuts to pull from. If so, continue.
+    if (lengthOfStartStack > 0) {
+      // this var donut is the value of the token being taken off of the stack. 
+      var donut = stacks[startStack][(lengthOfStartStack - 1)];
+      // this if checks to see whether the length of the stacked-on is < 0 OR|| if the value of the donut is less than 
+      // the value of the last number in the array lengthOfEndStack-1 of endStack of stacks. 
+      // we subtract 1 from the lengthOfEndStack because the index begins with 0 while the length of an array starts with 1
+      // So we have to subract one from the length else we would call on a numbe that isn't in the array, giving us "undefined"
```
Comment (pieroapretto) [inline apps/04TowersOfHanoi.js:44]: Great job leaving plenty of comments.

File: apps/04TowersOfHanoi.js [hunk 5]
```diff
@@ -19,27 +19,98 @@ function printStacks() {
...
+      if ((lengthOfEndStack === 0) || (donut < stacks[endStack][(lengthOfEndStack - 1)])){
```
Comment (pieroapretto) [inline apps/04TowersOfHanoi.js:45]: Nice use of the `||` operator in your `if` statement here.

File: homework/03LessonThree.js [hunk 6]
```diff
@@ -79,7 +79,8 @@ var problemSevenAnswer;
 // What is the ternary operator?  How does it differ from normal conditional statements?
 //
 // Your Answer Goes Here:
-//
+// Ternary operators use three operands and shorten code. Its simple says, 
+//if this is true do the first thing you come across after ? else do the second thing (after the :)
```
Comment (pieroapretto) [inline homework/03LessonThree.js:83]: correct.

File: homework/04LessonFour.js [hunk 7]
```diff
@@ -78,6 +90,7 @@ var nestedObjects = {
 };
 // update the starring property to have the value 'Tim Robbins'
 
+nestedObjects.users['bob roberts'].starring = 'Tim Robbins';
```
Comment (pieroapretto) [inline homework/04LessonFour.js:93]: nicely done

File: homework/02LessonTwo.js [hunk 8]
```diff
@@ -54,22 +68,25 @@ function lessThanOrEqualTo(parameter1, parameter2) {
...
-//
+// function whatMyNameIs(vars1, vars2, ...) {
+// tasks to do
+// }
 //
 // What is a parameter? What is an argument? Is there a difference between the two?
 //
 // Your Answer Goes Here:
-//
+// No there isn't a difference btw an argument and parameter, they can be set or changed by the user and used in the program.
```
Comment (pieroapretto) [inline homework/02LessonTwo.js:89]: Correctness = 100%!

File: apps/04TowersOfHanoi.js [hunk 9]
```diff
@@ -19,27 +19,98 @@ function printStacks() {
...
 
+var counter = 0;
+
+console.log("Welcome to the game TOWERS OF HANOI!!");
+console.log("To play simply type in the tower you'd like to move a token from");
+console.log("then type in the tower you'd like to place it.");
+console.log("Use only a, b or c and the token on top must always be smaller. Good Luck!");
+
 function getPrompt() {
+  console.log("Moves =" + " " + counter);
```
Comment (pieroapretto) [inline apps/04TowersOfHanoi.js:112]: This is very cool.

File: homework/01LessonOne.js [hunk 10]
```diff
@@ -106,7 +108,8 @@ var notNotTrue;
 // What is the difference between the "and" and the "or" operators? Explain why you would use each of them.
 //
 // Your Answer Goes Here:
-//
+// "and" returns the value on the left while "or" returns the one of the right. You would use each depending on which variable you need to compare.
```
Comment (pieroapretto) [inline homework/01LessonOne.js:111]: The 'and' operator evaluates the `truthiness` of the values on both values and only returns true if BOTH (left and right) are true.

The 'or' operator evaluates the `truthiness` of the values on both values and returns true if EITHER (left and right) are true.

File: apps/04TowersOfHanoi.js [hunk 11]
```diff
@@ -19,27 +19,98 @@ function printStacks() {
...
 
 function towersOfHanoi(startStack, endStack) {
-  // Your code here
 
+    // Will run to see if the move is legal before the move is made: 
+    if (isLegal(startStack, endStack)) 
+    {
+
+    // This runs movePiece 
+    movePiece(startStack, endStack);
```
Comment (jw56578) [inline apps/04TowersOfHanoi.js:91]: Make sure code inside a set of curly braces is indented to the right

File: apps/05Mastermind.js [hunk 12]
```diff
@@ -29,12 +29,62 @@ function getRandomInt(min, max) {
...
 }
 
-function generateHint() {
-  // your code here
+// Check to see if the guessed letters equal solution letters' location
+var correctLetterLocation = 0;
+//  Checks to see if any of the guessed letters equal the solutions letters in ANY location.
+var correctLetters  = 0;
+// 
+var hint = generateHint();
```
Comment (pieroapretto) [inline apps/05Mastermind.js]: Delete line 37. You're assigning var hint to a `generateSolution()` without passing in any parameters. This causes your app to run that function automatically without solution being passed through or defined.

File: apps/05Mastermind.js [hunk 13]
```diff
@@ -29,12 +29,62 @@ function getRandomInt(min, max) {
...
+    }
+  }
+  // console.log(solutionArray);
+  for (var i = 0; i < solutionArray.length; i++) {
+    // This targetIndex is a temporary container the holds the value of i in guessArray and evaluates the .indexOf it in solutionArray. 
+    var targetIndex = solutionArray.indexOf(guessArray[i]);
+    // Evaluates if the container targetIndex is > -1 (meaning, it exists in the solutionArray), if so it adds a count to correctLetters and sets the value to null.
+    if (targetIndex > -1 ){
+    correctLetters++;
+    solutionArray = null;
```
Comment (pieroapretto) [inline apps/05Mastermind.js]: The entire array should not nulled out after a match.

File: apps/05Mastermind.js [hunk 14]
```diff
@@ -29,12 +29,62 @@ function getRandomInt(min, max) {
   return Math.floor(Math.random() * (max - min)) + min;
 }
 
-function generateHint() {
-  // your code here
+// Check to see if the guessed letters equal solution letters' location
+var correctLetterLocation = 0;
```
Comment (pieroapretto) [inline apps/05Mastermind.js]: You've defined `correctLetterLocations` and `correctLetters` and set them to `0` outside of the `generateSolution()` function. When you run the game or a series of tests, these 2 variables never get reset to zero after the user makes an incorrect guess. Also, make sure you have these 2 variables spelled correctly throughout your game.

File: apps/06.1Checkers.js [hunk 15]
```diff
@@ -8,73 +8,185 @@ var rl = readline.createInterface({
...
+    // in var checker
+    this.board.grid[end[0]][end[1]] = checker;
+
+    var distance = Math.abs(start[0] - end[0]);
+    // Used the console.logs to figure out what the computer was calculating. Learned I needed to convert the string start and end to Number
+    // console.log("distance is " + distance);
+    // console.log("middle row is " + midRow);
+    // console.log("middle column is " + midColumn);   
+    // console.log(killPosition);   
+    if (distance === 2){
```
Comment (pieroapretto) [inline apps/06.1Checkers.js:157]: Nice use of an `if` statement here.

File: apps/06.1Checkers.js [hunk 16]
```diff
@@ -8,73 +8,185 @@ var rl = readline.createInterface({
...
+      var startY = Number(start[1]);
+      var endY = Number(end[1]);
+      // used the midpoint formula to find the average of x and y
+      var midRow = (startX + endX) / 2;
+      var midColumn = (startY + endY) / 2;
+      // created a new container to hold the value of the averages of x and y
+      var killPosition = [midRow, midColumn]; 
+      //  Then we send the value to the killChecker function; again as lon as the distance is 2
+      // Could come back later and make legal moves...
+      this.board.killChecker(killPosition);
```
Comment (pieroapretto) [inline apps/06.1Checkers.js:170]: ALL test passing! Good job leaving lots of comments too.

wizard

File: apps/05Mastermind.js [hunk 17]
```diff
@@ -29,18 +29,70 @@ function getRandomInt(min, max) {
...
+  }
+
+    // creates a variable to hold the correct Letter and locations to be used later.
+    return correctLetterLocations + "-" + correctLetters;
 }
 
 function mastermind(guess) {
-  // your code here
+  // This is a test line to make sure my function is working. it currently equals 'abcd'
+  solution = "abcd";
```
Comment (pieroapretto) [inline apps/05Mastermind.js:70]: You can remove the predefined solution here after your tests pass.

The game will generate one for you.

File: homework/06LessonSix.js [hunk 18]
```diff
@@ -37,33 +42,38 @@ var anotherPoint;
 // What is the difference between point (an object literal, or associative array)
 // and anotherPoint (a point object constructed from the Point class)?
 //
-//
+// point is a var we created literally or manually while anotherPoint was created using a class built to build objects. We passed
```
Comment (pieroapretto) [inline homework/06LessonSix.js:45]: Good answer.

File: homework/06LessonSix.js [hunk 19]
```diff
@@ -76,27 +86,31 @@ function ConferenceRoom() {
...
+conferenceRoom.enter(jen);
 
 // ****
 // Concept Checkpoint
 //
 // Write your answer in comments
 //
 // What is a method?
-//
+// A methods a similar to a function in that it does something with data and outputs something. A method is associated with
```
Comment (pieroapretto) [inline homework/06LessonSix.js:112]: Yup, a method is a function assigned to an object as a property of that object.

File: apps/08jquery-towers-of-hanoi/script.js [hunk 20]
```diff
@@ -1,5 +1,48 @@
...
+          // add in logic here to place the piece back on the stack it came from 
+        } 
+    }
+    checkForWIn();
+  
+// Build reset button
+// Build counter
+// Display counter
+
+} //<-- last } in handleWhatWasClicked
```
Comment (jw56578) [inline apps/08jquery-towers-of-hanoi/script.js]: Good helper comment

File: apps/08jquery-towers-of-hanoi/script.js [hunk 21]
```diff
@@ -105,39 +113,39 @@ $(document).on('mousemove', function(e){
...
+}  // <--- last } in generateBlocks  
 
-  // john@digitaldownbeat.com
-  // added this line as a test
-
-  // Line 128, added in the data attribute to each new div. I can use this value to compare to find if the move is legal.
\ No newline at end of file
+// returns the value of the number needed to win and pushes the value to a div.
+function pushMinimum() {
+  $("#block-Counter").text("Min # of moves " + (Math.pow(2, toWinValue) - 1));
```
Comment (pieroapretto) [inline apps/08jquery-towers-of-hanoi/script.js]: Sweet sassy molassy! Nice function.

File: apps/08jquery-towers-of-hanoi/script.js [hunk 22]
```diff
@@ -59,36 +68,35 @@ function handleWhatWasClicked(){
...
 
-// Built separately to easily change later
 function checkForWIn() {
   // Creates vars to measure the amount of data in the arrays.
   var lengthOfLastStack = $('[data-stack = "3"]').children().length;
   var lengthOfSecondStack = $('[data-stack = "2"]').children().length;
 
   // compares the values of the arrays to 4 because we have 4 game pieces.
-  if (lengthOfLastStack === 4 || lengthOfSecondStack === 4) {
+  if (lengthOfLastStack === toWinValue || lengthOfSecondStack === toWinValue) {
```
Comment (pieroapretto) [inline apps/08jquery-towers-of-hanoi/script.js]: Smart way to handle checkForWin() based on user input!

File: apps/07jquery-tic-tac-toe/script.js [hunk 23]
```diff
@@ -2,4 +2,100 @@
...
+    if($(this).text() === "X" || $(this).text() === "O" ) {
+      // prevent move
+      alert("Choose another spot!");
+    }
+    // if the spot is :empty go ahead and run this part fo the function
+      else {
+        // will use the .text() method to put the value of playerTurn into .this instance of data-cell 
+        $(this).text(playerTurn); 
+        //  after making a move this if statment runs the three functions inside to get a true/false value
+        if (horizontalWin() || verticalWin() || diagonalWin() === true){
```
Comment (pieroapretto) [inline apps/07jquery-tic-tac-toe/script.js:23]: Clean up if statement -> `if (horizontalWin() || verticalWin() || diagonalWin()){ ..`

**Pull request closed without merging**